Hmusum/support deleting local and remote session in one go

hmusum opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Commit:
Add test simulating delay between deleting local and remote sessions

Commit:
Use a Runnable for advancing clock when testing maintainer

Commit:
Add new method for deleting expired config sessions

Add new method that will delete both remote and local sessions at
the same time, if they have expired. Controlled  by feature flag.
Also added test variant that uses new method, showing that a scenario
works with the new method but not with the old.

jonmv commented on configserver/src/test/java/com/yahoo/vespa/config/server/maintenance/SessionsMaintainerTest.java:
```diff
@@ -189,6 +186,64 @@ public void testDeletingInactiveSessions3() throws IOException {
+        // All sessions except session id 3 should be removed after maintainer has run
+        maintainer.maintain(Duration.ofSeconds(5));
+        assertNumberOfLocalSessions(2);
+        assertNumberOfRemoteSessions(2);
```
Why 2, then?

jonmv commented on configserver/src/test/java/com/yahoo/vespa/config/server/maintenance/SessionsMaintainerTest.java:
```diff
@@ -210,16 +210,19 @@ public void testDeletionWithDelay() {
+
+        // Sessions 3 and 4 should be left after maintainer has run, 2 should be deleted
+        // Test this with advancing clock 5 seconds between deletion of local and remote sessions
+        maintainer.maintain(() -> clock.advance(Duration.ofSeconds(5)));
```
👍

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+                if (deleted >= deleteMax)
+                    return;
+            }
+        } catch (Throwable e) { // Make sure to catch here, to avoid executor just dying in case of issues ...
```
Should the try-catch be per session, to avoid one failing session blocking deletion of others? Alternatively, use a random or time-dependent subset of sessions—perhaps the first 0.05, but in reverse creation time order?

jonmv replied:
Ah, they're in somewhat random order already, from being a hash set of longs ... 
and using reverse creation order potentially leaves older sessions hanging, of course. 
So some true randomness, then, or moving the try-catch down a step.

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+            for (Long sessionId : sessions) {
+                // Skip sessions newly added (we might have a session in the file system, but not in ZooKeeper,
+                // we will exclude these)
+                if (newSessions.contains(sessionId)) continue;
```
Perhaps simpler to `sessions.removeAll(newSessions)` outside the loop?

jonmv replied:
Well, this fits with the below `continue`, so perhaps not ...

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+                        ", created: " + createTime + ", status " + status + ", can be deleted: " + canBeDeleted(sessionId, status) +
+                        ", hasExpired: " + hasExpired);
+                if (hasExpired && canBeDeleted(sessionId, status)) {
+                    log.log(Level.FINE, () -> "expired: " + hasExpired + ", can be deleted: " + canBeDeleted(sessionId, status));
```
I'd think `canBeDeleted` is always true here, unless there's a race, which we don't want?

jonmv replied:
Looks like a redundant log line, with the logging a few lines up?

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+                if (hasExpired && canBeDeleted(sessionId, status)) {
+                    log.log(Level.FINE, () -> "expired: " + hasExpired + ", can be deleted: " + canBeDeleted(sessionId, status));
+                    deleteLocalSession(sessionId);
+                } else if (createTime.plus(Duration.ofDays(1)).isBefore(clock.instant())) {
```
Could expiration time ever exceed 1 day? If it does, this becomes a bit weird—it's not "expired", but we attempt forceful deletion anyway.

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+                    log.log(Level.FINE, () -> "more than 1 day old: " + sessionId);
+                    try {
+                        localSession = getSessionFromFile(sessionId);
+                    } catch (Exception e) {
+                        log.log(Level.FINE, () -> "could not get session from file: " + sessionId + ": " + e.getMessage());
+                        continue;
+                    }
+                    Optional<ApplicationId> applicationId = localSession.getOptionalApplicationId();
+                    if (applicationId.isEmpty()) continue;
```
Are these expected states?

jonmv reviewed (commented):
I like this a lot better. I believe there could still be cases where we delete a remote session, but not the local one, but that it would only happen if something else was wrong (`UNKNOWN` session status, or inconsistent state (`ACTION` session which isn't "the active session" of an application).

Commit:
Some fixes after feedback in PR

Remove new sessions (should be ignored) at once.
Move try-catch to be per session.
Split out a method fot getting local sessions from file	system.
Some minor changes to logging.

Commit:
Simplify logic a bit and add comments for corner cases

hmusum commented:
I think I have addressed most of you comments, there are some corner cases and perhaps excessive checking before deleting in the code, but rather safe than sorry...

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,97 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+                if (session == null)
+                    session = new RemoteSession(tenantName, sessionId, createSessionZooKeeperClient(sessionId));
+
+                Session.Status status = session.getStatus();
+                boolean activeForApplication = sessionIsActiveForApplication.test(session);
+                if (status == ACTIVATE && activeForApplication) continue;
+
+                Instant createTime = session.getCreateTime();
+                boolean hasExpired = hasExpired(createTime);
+                if ( ! hasExpired) continue;
+
+                log.log(Level.FINE, () -> "Remote session " + sessionId + " for " + tenantName + " has expired, deleting it");
+                deleteRemoteSessionFromZooKeeper(session);
+                deleted++;
```
So I guess this is racy: status and active-session are written in one transaction, but here we read them one after the other, so we could potentially see `PREPARED` and `true`, and then go on to delete the remote session, if it had spent a long time in prepare (not impossible when a session fails to activate because of host provisioning). 
I guess we could handle that by simply re-reading status if `activeForApplication` but `status != ACTIVATE`, or re-reading active-session in the other case? If they still disagree, we should probably warn about it?

jonmv replied:
I wonder if we should also bail out if they disagree, unless the session `isOldAndCanBeDeleted(...)`, like for local sessions?

jonmv replied:
Hmm, we could also read the session status as `PREPARED`, and active as `false`, but then the session is activated while we delete it. I don't remember without looking at the code, but I believe those writes are protected by a lock. Alternatively, we could do the deletion with a zknode version condition ... which I'm not sure we expose through our own APIs >_<

hmusum replied:
Will do this while holding session lock instead

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java:
```diff
@@ -682,6 +682,97 @@ private boolean sessionLifeTimeElapsed(Instant created) {
+
+                log.log(Level.FINE, () -> "Expired local session is candidate for deletion: " + sessionId +
+                        ", created: " + createTime + ", status " + status + ", can be deleted: " + canBeDeleted(sessionId, status));
+                // Need to check if it can be deleted, as we might have a (local) session in UNKNOWN
+                // state that we should not delete
```
Do you remember how we could end up in this state?

Commit:
Take session lock

hmusum commented:
@jonmv please take a look again

jonmv reviewed (approved):
Wohoo!
